datafusion/expr/src/type_coercion/binary.rs

```diff
@@ -87,9 +87,11 @@ fn signature(lhs: &DataType, op: &Operator, rhs: &DataType) -> Result<Signature>
         }
         And | Or => if matches!((lhs, rhs), (Boolean | Null, Boolean | Null)) {
             // Logical binary boolean operators can only be evaluated for
-            // boolean or null arguments.                   
+            // boolean or null arguments.                 
             Ok(Signature::uniform(DataType::Boolean))
-        } else {
+        } else if matches!((lhs, rhs), (List(_), List(_))) {
+            Ok(Signature::uniform(lhs.clone()))
```

Comment:
Don't we need to make sure left and right `List` have same element type?

Reply:
In the long term, we should have the same signature as BuiltinFunction::ArrayIntersect

Reply:
I reuse `array_coercion` to resolve the data type~